tests/test_energy_overflows.py

```diff
@@ -0,0 +1,126 @@
+import numpy as np
+import pytest
+from common import GradientTest, prepare_nb_system, prepare_water_system
+
+from timemachine.integrator import FIXED_TO_FLOAT
+from timemachine.potentials import NonbondedAllPairs, NonbondedPairListNegated
+
+pytestmark = [pytest.mark.memcheck]
+
+
+@pytest.mark.parametrize("precision,rtol,atol", [(np.float64, 1e-8, 1e-8), (np.float32, 1e-4, 5e-4)])
+def test_energy_overflow_due_to_clashes(precision, rtol, atol):
+    """This test validates that if two particles overlap, the Nonbonded potential will return an inf energy"""
+    np.random.seed(2023)
+
+    N = 2
+    D = 3
+
+    x0 = np.random.rand(N, D).astype(dtype=np.float64) * 2
+
+    # Have no exclusions, only care about intermolecular energies, as exclusions should be removed
+    E = 0
+
+    box = np.eye(3) * 100.0
+
+    params, potential = prepare_nb_system(x0, E, p_scale=1.0, cutoff=1.0)
+    params[:, 3] = 0.0  # Remove 4d offset
+
+    gpu_pot = potential.to_gpu(precision)
+
+    gpu_unbound = gpu_pot.unbound_impl
+    gpu_bound = gpu_pot.bind(params).bound_impl
+
+    reference_energy = potential(x0, params, box)
+    _, _, test_energy = gpu_unbound.execute_selective(x0, params, box, False, False, True)
+    np.testing.assert_allclose(reference_energy, test_energy, rtol=rtol, atol=atol)
+
+    # Verify that there was no overflow
+    _, overflows = gpu_bound.execute_fixed(x0, box)
+    assert overflows == 0
+
+    # Set the two particles to be overlapping
+    x0[0, :] = x0[1, :]
+
+    # Note that the reference returns a nan, while the unbound returns an inf
+    reference_energy = potential(x0, params, box)
+    assert not np.isfinite(reference_energy)
+    _, _, test_energy = gpu_unbound.execute_selective(x0, params, box, False, False, True)
+    assert not np.isfinite(test_energy)
+
+    # Verify that there was an overflow that wasn't cancelled out
+    _, overflows = gpu_bound.execute_fixed(x0, box)
+    assert overflows != 0
+
+
+@pytest.mark.parametrize("precision,rtol,atol", [(np.float64, 1e-8, 1e-8), (np.float32, 1e-4, 5e-4)])
+def test_energy_overflow_cancelled_by_exclusions(precision, rtol, atol):
+    """This test validates that if two particles overlap, the Nonbonded potential will return an inf energy"""
+    np.random.seed(2023)
+
+    N = 9  # Multiple of three for waters
+    D = 3
+
+    x0 = GradientTest().get_water_coords(D)[:N]
```

Comment:
Does this actually return a clashy system?

Reply:
It does not, the intention with this test is to verify that the 'clashes' in the AllPairs and PairListNegated cancel out.